### DISCUSSION TITLE
NextJs app api with 500 on Vercel platform

### DISCUSSION THREAD

**Discussion by johnfung95** [vercel/next.js#55214] [Help]
### Summary

I have deployed a NextJs app onto the Vercel platform. However, the api in the app triggers the 500 internal server error. The app is supposed to connect to a backend SQL server through prisma. While I could do it in development mode, I could not overcome the error in the production stage. I am still confused as to what causes the error. I've attached the 2 api logs on found on Vercel. 

Please help to provide some insights on the matter. Thank you very much for any assistance.

### Additional information

```js
api1: /api/details
`Unhandled Promise Rejection 	{"errorType":"Runtime.UnhandledPromiseRejection","errorMessage":"PrismaClientInitializationError: Timed out fetching a new connection from the connection pool. More info: http://pris.ly/d/connection-pool (Current connection pool timeout: 10, connection limit: 5)","reason":{"errorType":"PrismaClientInitializationError","errorMessage":"Timed out fetching a new connection from the connection pool. More info: http://pris.ly/d/connection-pool (Current connection pool timeout: 10, connection limit: 5)","name":"PrismaClientInitializationError","clientVersion":"4.16.2","errorCode":"P2024","stack":["PrismaClientInitializationError: Timed out fetching a new connection from the connection pool. More info: http://pris.ly/d/connection-pool (Current connection pool timeout: 10, connection limit: 5)","    at r (/var/task/node_modules/@prisma/client/runtime/library.js:108:2574)"]},"promise":{},"stack":["Runtime.UnhandledPromiseRejection: PrismaClientInitializationError: Timed out fetching a new connection from the connection pool. More info: http://pris.ly/d/connection-pool (Current connection pool timeout: 10, connection limit: 5)","    at process.<anonymous> (file:///var/runtime/index.mjs:1250:17)","    at process.emit (node:events:526:35)","    at emit (node:internal/process/promises:149:20)","    at processPromiseRejections (node:internal/process/promises:283:27)","    at process.processTicksAndRejections (node:internal/process/task_queues:96:32)"]}
Unknown application error occurred
Runtime.Unknown`

api2: /api/get-groups
`PrismaClientInitializationError: 
Invalid `prisma.apigroup.findMany()` invocation:

Timed out fetching a new connection from the connection pool. More info: http://pris.ly/d/connection-pool (Current connection pool timeout: 10, connection limit: 5)
    at Rn.handleRequestError (/var/task/node_modules/@prisma/client/runtime/library.js:174:7601)
    at Rn.handleAndLogRequestError (/var/task/node_modules/@prisma/client/runtime/library.js:174:6754)
    at /var/task/node_modules/@prisma/client/runtime/library.js:177:3010
    at async /var/task/node_modules/@prisma/client/runtime/library.js:177:3225
    at async t._executeRequest (/var/task/node_modules/@prisma/client/runtime/library.js:177:11403)
    at async GET (/var/task/.next/server/app/api/get-groups/route.js:72:21)
    at async /var/task/.next/server/chunks/601.js:2903:37 {
  clientVersion: '4.16.2',
  errorCode: undefined
}
Unhandled Promise Rejection 	{"errorType":"Runtime.UnhandledPromiseRejection","errorMessage":"PrismaClientInitializationError: Timed out fetching a new connection from the connection pool. More info: http://pris.ly/d/connection-pool (Current connection pool timeout: 10, connection limit: 5)","reason":{"errorType":"PrismaClientInitializationError","errorMessage":"Timed out fetching a new connection from the connection pool. More info: http://pris.ly/d/connection-pool (Current connection pool timeout: 10, connection limit: 5)","name":"PrismaClientInitializationError","clientVersion":"4.16.2","errorCode":"P2024","stack":["PrismaClientInitializationError: Timed out fetching a new connection from the connection pool. More info: http://pris.ly/d/connection-pool (Current connection pool timeout: 10, connection limit: 5)","    at r (/var/task/node_modules/@prisma/client/runtime/library.js:108:2574)"]},"promise":{},"stack":["Runtime.UnhandledPromiseRejection: PrismaClientInitializationError: Timed out fetching a new connection from the connection pool. More info: http://pris.ly/d/connection-pool (Current connection pool timeout: 10, connection limit: 5)","    at process.<anonymous> (file:///var/runtime/index.mjs:1250:17)","    at process.emit (node:events:526:35)","    at emit (node:internal/process/promises:149:20)","    at processPromiseRejections (node:internal/process/promises:283:27)","    at process.processTicksAndRejections (node:internal/process/task_queues:96:32)"]}
Unknown application error occurred
Runtime.Unknown`
```

**Comment by devjiwonchoi** [comment]
Try to deploy the `api` route first then deploy the corresponding sources.

**Reply by johnfung95** [reply]
Thanks for your response. How exactly do you mean to deploy the routes first? Can it be configured with a file?

**Reply by devjiwonchoi** [reply]
Since `route.js` doesn't run on the build time, I've experienced that the page component that depends on that route.js breaks due to missing the data. I was assuming that this is a similar case. The solution for that was to push the `route.js` file only first, then push the rest of the commits.

**Reply by johnfung95** [reply]
Thanks. I will give it a shot, but as far as I see from the error messages. It is the Prisma initialization problem. Therefore I am more concerned about the use of Prisma and Vercel deployment. Maybe there are bugs or my deployment method was incorrect?

**Reply by devjiwonchoi** [reply]
Maybe, and if so, we need to dig down to the codes I guess, and probably need a repro.

**Reply by johnfung95** [reply]
I tried to push the api routes first but the result is the same. I have created a replicated version for inspection:
https://github.com/johnfung95/test_vercel_dashboard.git

Basically I am connecting to a sqlserver through prisma for retrieving my data. Again, everything works in localhost, just not the production stage.
In this replicated version, I only use the Dashboard component to get the details and it will only display the results in the console.

My connection string is as follows:
sqlserver://xxx.xx.xx.x:1433;database=Dashboard;user=xx;password=xxxxxxxxx;encrypt=true;trustServerCertificate=true;connectTimeout=0;

The error codes I encounter is either 504 or 500.

**Reply by Lucid-149** [reply]
did you get a solution?

**Reply by cammycurry** [reply]
any solution?

**Comment by johnfung95** [comment]
oh, I forgot to update the status, my bad. I resolved the issue by hosting my website with IIS instead. I realized it was probably some IP whitelisting or cert-trusting issues with Vercel. Therefore I chose to go an easier way by not using Vercel for deployment.

**Reply by cammycurry** [reply]
I ended up solving by increasing the pooling timeout via the database url... 

https://www.prisma.io/docs/orm/prisma-client/setup-and-configuration/databases-connections/connection-pool

postgresql://johndoe:mypassword@localhost:5432/mydb?connection_limit=5&pool_timeout=2

**Reply by johnfung95** [reply]
@cammycurry  Good for you. I considered that possibility too. But in my case, my database is placed in a server where IP whitelisting is required for incoming traffic, so it does not work for me. Still, I am glad it helps in your case.
